cmake/Modules/FindFLANN.cmake

```diff
@@ -1,63 +1,145 @@
-###############################################################################
-# Find FLANN
+#.rst:
+# FindFLANN
+# --------
 #
-# This sets the following variables:
-# FLANN_FOUND - True if FLANN was found.
-# FLANN_INCLUDE_DIRS - Directories containing the FLANN include files.
-# FLANN_LIBRARIES - Libraries needed to use FLANN.
-# FLANN_DEFINITIONS - Compiler flags for FLANN.
-# If FLANN_USE_STATIC is specified and then look for static libraries ONLY else
-# look for shared ones
+# Try to find FLANN library and include files.
+#
+# IMPORTED Targets
+# ^^^^^^^^^^^^^^^^
+#
+# This module defines the :prop_tgt:`IMPORTED` targets:
+#
+# ``FLANN::FLANN``
+#  Defined if the system has FLANN.
+#
+# Result Variables
+# ^^^^^^^^^^^^^^^^
+#
+# This module sets the following variables:
+#
+# ::
+#
+#   FLANN_FOUND               True in case FLANN is found, otherwise false
+#   FLANN_DEFINITIONS         Compiler flags for FLANN.
+#   FLANN_INCLUDE_DIR         Location of FLANN header files
+#   FLANN_INCLUDE_DIRS        Location of FLANN header files (including dependencies)
+#   FLANN_LIBRARY             FLANN release library
+#   FLANN_LIBRARY_DEBUG       FLANN debug library
+#   FLANN_LIBRARIES           FLANN release and debug library
+#
+# Example usage
+# ^^^^^^^^^^^^^
+#
+# ::
+#
+#     find_package(FLANN REQUIRED)
+#
+#     add_executable(foo foo.cc)
+#     target_link_libraries(foo FLANN::FLANN)
+#
+
+find_package(PkgConfig)
+if(FLANN_FIND_VERSION)
+    pkg_check_modules(PC_FLANN flann>=${FLANN_FIND_VERSION})
+else()
+    pkg_check_modules(PC_FLANN flann)
+endif()
+
+set(FLANN_DEFINITIONS ${PC_FLANN_CFLAGS_OTHER})
 
 if(FLANN_USE_STATIC)
   set(FLANN_RELEASE_NAME flann_cpp_s)
   set(FLANN_DEBUG_NAME flann_cpp_s-gd)
+  set(FLANN_DEFINITIONS ${FLANN_DEFINITIONS} "-DFLANN_STATIC")
 else()
   set(FLANN_RELEASE_NAME flann_cpp)
   set(FLANN_DEBUG_NAME flann_cpp-gd)
 endif()
 
-find_package(PkgConfig QUIET)
-if(FLANN_FIND_VERSION)
-    pkg_check_modules(FLANN flann>=${FLANN_FIND_VERSION})
-else()
-    pkg_check_modules(FLANN flann)
-endif()
-
-if(NOT FLANN_FOUND)
-    find_path(FLANN_INCLUDE_DIR flann/flann.hpp
-              HINTS "${FLANN_ROOT}" "$ENV{FLANN_ROOT}"
-              PATHS "$ENV{PROGRAMFILES}/flann" "$ENV{PROGRAMW6432}/flann"
-              PATH_SUFFIXES include)
-
-    find_library(FLANN_LIBRARY
-                 NAMES ${FLANN_RELEASE_NAME}
-                 HINTS "${FLANN_ROOT}" "$ENV{FLANN_ROOT}"
-                 PATHS "$ENV{PROGRAMFILES}/flann" "$ENV{PROGRAMW6432}/flann"
-                 PATH_SUFFIXES lib)
-
-    find_library(FLANN_LIBRARY_DEBUG
-                 NAMES ${FLANN_DEBUG_NAME} ${FLANN_RELEASE_NAME}
-                 HINTS "${FLANN_ROOT}" "$ENV{FLANN_ROOT}"
-                 PATHS "$ENV{PROGRAMFILES}/flann" "$ENV{PROGRAMW6432}/flann"
-                 PATH_SUFFIXES lib debug/lib)
-
-    if(NOT FLANN_LIBRARY_DEBUG)
-      set(FLANN_LIBRARY_DEBUG ${FLANN_LIBRARY})
-    endif()
+find_path(FLANN_INCLUDE_DIR
```

Comment:
Well, as it is, this will fix bug #2802. This can be circumvented if we are able to generate full paths to the libraries from the information pkg-config provides. In case it's a couple of libs in a single folder this is trivial, but if it's a couple of libs from multiple folders it becomes more troublesome.

The need for having absolute paths to libs comes from CMake convention. Quoting from the [CMake docs page on link_directories](https://cmake.org/cmake/help/v3.0/command/link_directories.html).
> Note that this command is rarely necessary. Library locations returned by find_package() and find_library() are absolute paths. Pass these absolute library file paths directly to the target_link_libraries() command. CMake will ensure the linker finds them.

This expectation is so prevalent that there's no target interface properties regarding the lib directories. [Here's](https://cmake.org/cmake/help/v3.0/manual/cmake-properties.7.html?highlight=target%20properties#properties-on-targets) the extensive list. This becomes a problem for our downstream projects which are not explicitly using FLANN. It forces us to add an unconditional `link_libraries(FLANN_LIBRARY_DIRS)` during PCLConfig.cmake which is a little less clean.

There's also the option of setting linker flags and passing `-L` for nix systems and whatever is the equivalent on Windows. It feels hacky to me though.

Reply:
You are right, this is documented for CMake here: https://cmake.org/cmake/help/latest/manual/cmake-developer.7.html#a-sample-find-module

sorry for the confusion.

Reply:
So nothing to change here and I can remove `link_libraries(FLANN_LIBRARY_DIRS)`?

Reply:
Yes :).